### ISSUE TITLE
Drop Python 2.5 support after Twisted 12.1 release

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5553]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5553|
|Type|task|
|Created|2012-03-14 21:51:37Z|
|Branch|https://github.com/twisted/twisted/tree/drop-25-support-5553-2|

In order to make it easier to support Python 3, but also to reduce maintenance burden, we should drop support for Python 2.5 after our next release. Asking on the mailing list didn't bring up anyone who would have issues with this, and we'll still support it in 12.1.

<details><summary>Searchable metadata</summary>

```
trac-id__5553 5553
type__task task
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_drop_25_support_5553_2 branches/drop-25-support-5553-2
branch_author__vperic vperic
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1331761897000000 1331761897000000
changetime__1343986768000000 1343986768000000
version__None None
owner__vperic vperic
cc__zooko@... cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_1"></a>|@zooko commented|
|-|-|

Here's the analogous ticket for Tahoe-LAFS to drop support for Python ≤ 2.5, which we've been planning to do "after the next release", for the last several releases:

https://tahoe-lafs.org/trac/tahoe-lafs/ticket/1658

You might be interested in my notes about which major operating systems come with which versions of Python.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_3"></a>|@vperic commented|
|-|-|

(In [34599]) Create branch drop-25-support-5553

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_4"></a>|@vperic commented|
|-|-|

There we go.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set owner to @vperic|
|-|-|

Thanks for working on this. I'm not sure if this ticket is about updating the documentation, like you did in the branch, or also includes actually removing workarounds etc for Python 2.5. There are still quite some references. Could you also open several new ticket for removing those references (dividing it into separate tickets reduces the amount of work for a contributor and reviewer) and put this back up for review?

And maybe you can also mention in the .removal file that Python 2.5 is in security fix mode now and was released in xxxx (? ages ago) to emphasize people should upgrade their python if they still use 2.5?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

The references I referred to in the previous comment:

```
./twisted/trial/test/test_assertions.py:            # Python 2.5 and earlier, stacklevel of 0 resulted in a warning
./twisted/__init__.py:    raise RuntimeError("Twisted requires Python 2.5 or later.")
./twisted/web/xmlrpc.py:            raise RuntimeError("useDateTime requires Python 2.5 or later.")
./twisted/web/xmlrpc.py:            raise RuntimeError("useDateTime requires Python 2.5 or later.")
./twisted/web/_flatten.py:            # In Python 2.5, after an exception, a generator's gi_frame is
./twisted/topfiles/setup.py:    print >>sys.stderr, "You must use at least Python 2.5 for Twisted"
./twisted/python/filepath.py:            # Under Python 2.5 on Windows, WindowsError has a winerror
./twisted/python/filepath.py:            # ERROR_PATH_NOT_FOUND.  However, in Python 2.5,
./twisted/python/filepath.py:    precision, call C{os.stat_float_times(True)}, or use Python 2.5.
./twisted/internet/defer.py:    requirement of Python 2.5, consider the use of L{inlineCallbacks} instead,
./twisted/internet/defer.py:    regular sequential function. This function uses features of Python 2.5
./twisted/internet/test/test_sigchld.py:    # random), Python 2.5 has set_wakeup_fd but not siginterrupt.
./twisted/internet/test/test_inlinecb.py:Conditional import of C{inlinecb_tests} for Python 2.5 and greater.
./twisted/internet/selectreactor.py:            # subclass, but on Python 2.5 and earlier it is not.)
./twisted/conch/checkers.py:    # Python 2.5 got spwd to interface with shadow passwords
./twisted/test/generator_failure_tests.py:Python 2.5+ test cases for failures thrown into generators.
./twisted/words/protocols/jabber/xmpp_stringprep.py:    # We require Unicode version 3.2. Python 2.5 and later provide this as
./twisted/application/app.py:                # Python 2.5 and above supports a stream attribute
```

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_7"></a>|@vperic removed owner|
|-|-|

I opened ticket [#5715](https://github.com/twisted/twisted/issues/5715) for the removal of code references you mention; just one ticket because the same tickets for 2.3 and 2.4 are about the same size (also marked it with the easy keyword). I also changed the topfile to now read "Python 2.5 is no longer supported. Python 2.5 was released in 2006 and will no longer receive even security fixes; users are encouraged to upgrade."

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> And maybe you can also mention in the .removal file that Python 2.5 is in security fix mode now and was released in xxxx (? ages ago) to emphasize people should upgrade their python if they still use 2.5?

Why are we telling people about CPython's release timeline?  We should leave that up to python.org.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

Replying to [exarkun](#note_8):
> Why are we telling people about CPython's release timeline?  We should leave that up to python.org.

I was trying to do the users of twisted a favor so they can assess more quickly why this decision was made, like mentioning the version nr of Twisted in .removal files.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra set owner to @vperic|
|-|-|

Replying to [vperic](#note_7):
> I opened ticket [#5715](https://github.com/twisted/twisted/issues/5715) for the removal of code references you mention; just one ticket because the same tickets for 2.3 and 2.4 are about the same size (also marked it with the easy keyword). I also changed the topfile to now read "Python 2.5 is no longer supported. Python 2.5 was released in 2006 and will no longer receive even security fixes; users are encouraged to upgrade."

If you can [comment:8 simplify] the topfile a little to something like "Python 2.5 is no longer supported.", then merge after a successful buildbot run.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_11"></a>|@vperic commented|
|-|-|

Simplified the topfile back, running the tests now: [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/drop-25-support-5553).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra commented|
|-|-|

vperic, could you merge this now that we started adding Python 2.6+ code in the trunk? Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_13"></a>|@vperic commented|
|-|-|

(In [35031]) Create branch drop-25-support-5553-2

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_14"></a>|@vperic commented|
|-|-|

I ordered a test run for the new branch, [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/drop-25-support-5553-2).

**Comment by twisted-trac** [comment]
|[<img alt="vperic's avatar" src="https://avatars.githubusercontent.com/u/688281?s=50" width="50" height="50">](https://github.com/vperic)<a name="note_15"></a>|@vperic set status to `closed`|
|-|-|

(In [35033]) Merge drop-25-support-5553-2: Remove support for Python 2.5

Author: vperic
Reviewer: thijs
Fixes: [#5553](https://github.com/twisted/twisted/issues/5553)

Python 2.5 is no longer supported, modify the appropriate setup files
and documentation.
